talk/morelanguage/ranges.tex

```diff
@@ -0,0 +1,91 @@
+    \item mapping the different types of iterators
+    \end{itemize}
+  \end{exampleblock}
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitlecpp[20]{Views}
+  \begin{block}{The view concept}
+    \begin{itemize}
+    \item A \mintinline{cpp}{view} is a non owning object build from a range or another view after applying some range adaptor(s)
+    \item Its time to copy, move, assign is constant
+    \item Range adaptors can be applied with \mintinline{cpp}{|} and are lazy
+    \item They live in the \mintinline{cpp}{std::views} namespace
+    \end{itemize}
+  \end{block}
+  \begin{exampleblock}{Example}
+    { \scriptsize
+      \begin{cppcode*}{gobble=4}
+        auto const numbers = std::views::iota(0, 6);
+        auto even = [](int i) { return 0 == i % 2; };
+        auto square = [](int i) { return i * i; };
+        auto results = numbers | std::views::filter(even)
+                               | std::views::transform(square);
+        for (auto a : results) { ... } // 0, 4, 16
+      \end{cppcode*}
+    }
+  \end{exampleblock}  
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitlecpp[20]{Many range adaptors are provided by the STL}
+  \begin{block}{Useful subset}
+    \begin{description}[keys/values]
+    \item[all] a view on all elements
+    \item[filter] applies a filter
+    \item[transform] applies a transformation
+    \item[take] takes only first n elements
+    \item[drop] skips first n elements
+    \item[join/split] joins ranges or splits into subranges
+    \item[reverse] reverses view order
+    \item[elements] for tuple views, extract nth elements of each tuple
+    \item[keys/values] for pair like views, takes 1st/2nd element of each pair
+    \end{description}
+  \end{block}
+  Remember you can combine them via (\mintinline{cpp}{|})
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitlecpp[23]{Lazyness and functional programming}
+  \begin{exampleblock}{\texttt{view}s are lazy}
+    Computation is only trigerred when iterating
+    { \scriptsize
+      \begin{cppcode*}{gobble=4}
+        // print first 20 prime numbers above 1000000
+        for (int i: std::views::iota(1000000) | std::views::filter(odd) 
+                                              | std::views::filter(isPrime) 
+                                              | std::views::take(20)) {
+          std::cout << i << " ";  
+        }
+      \end{cppcode*}
+    }
+  \end{exampleblock}  
+  \begin{block}{\texttt{view}s allow functional programming with the STL}
+    { \scriptsize
+      \begin{cppcode*}{gobble=4}
+        auto const ints = std::views::iota(0, 6);
+        auto even = [](int i) { return 0 == i % 2; };
+        auto square = [](int i) { return i * i; };
+        auto results = numbers | std::views::filter(even)
+                               | std::views::transform(square);
+        for (auto a : results) { ... } // 0, 4, 16
+      \end{cppcode*}
+     }
+  \end{block}
```

Comment:
Why is this box identical to the example on the previous slide? Did you have in mind to mention something specific about this code?